compiler/sem/semexprs.nim

```diff
@@ -2932,26 +2940,88 @@ proc semTupleConstr(c: PContext, n: PNode, flags: TExprFlags): PNode =
 include semobjconstr
 
 proc semBlock(c: PContext, n: PNode; flags: TExprFlags): PNode =
-  result = n
-  inc(c.p.nestedBlockCounter)
-  checkSonsLen(n, 2, c.config)
-  openScope(c) # BUGFIX: label is in the scope of block!
-  if n[0].kind != nkEmpty:
-    var labl = newSymG(skLabel, n[0], c)
-    if sfGenSym notin labl.flags:
-      addDecl(c, labl)
-    elif labl.owner == nil:
-      labl.owner = c.p.owner
-    n[0] = newSymNode(labl, n[0].info)
-    suggestSym(c.graph, n[0].info, labl, c.graph.usageSym)
-    styleCheckDef(c.config, labl)
-    onDef(n[0].info, labl)
-  n[1] = semExpr(c, n[1], flags)
-  n.typ = n[1].typ
-  if isEmptyType(n.typ): n.transitionSonsKind(nkBlockStmt)
-  else: n.transitionSonsKind(nkBlockExpr)
-  closeScope(c)
-  dec(c.p.nestedBlockCounter)
+  addInNimDebugUtils(c.config, "semBlock", n, result, flags)
+
+  assert n != nil
+
+  case n.kind
+  of nkError:
+    result = n
+  of nkBlockExpr, nkBlockStmt:
+    checkSonsLen(n, 2, c.config)
+    inc(c.p.nestedBlockCounter)
+    openScope(c) # BUGFIX: label is in the scope of block!
+
+    # handle the label
+    let 
+      givenLabl = n[0]
+      lablRes =
+        case givenLabl.kind
+        of nkEmpty, nkError:
+          givenLabl
+        of nkIdent, nkSym, nkAccQuoted:
+          let labl = newSymG(skLabel, givenLabl, c)
+          
+          if sfGenSym notin labl.flags:
+            addDecl(c, labl)
+          elif labl.owner == nil:
+            labl.owner = c.p.owner
+          
+          let lablNode = newSymNode2(labl, givenLabl.info)
+
+          suggestSym(c.graph, lablNode.info, labl, c.graph.usageSym)
+          styleCheckDef(c.config, labl)
+          onDef(lablNode.info, labl)
+
+          lablNode
+        else:
+          c.config.newError(givenLabl,
+                            reportAst(rsemIdentExpectedInExpr, givenLabl))
+      bodyRes = semExpr(c, n[1], flags)
+
+    result = copyNode(n)
+    result.flags = n.flags # xxx: we should be able to perserve flags, this
+                           #      code didn't used to do a copyNode, it changed
+                           #      `n` directly
+    result.add lablRes
+    result.add bodyRes
+    
+    if bodyRes.kind != nkError:
+      result.typ = bodyRes.typ
+
+    if isEmptyType(result.typ):
+      result.transitionSonsKind(nkBlockStmt)
+    else:
+      result.transitionSonsKind(nkBlockExpr)
+
+    if lablRes.kind == nkError or bodyRes.kind == nkError:
+      result = c.config.wrapError(result)
```

Comment:
@Clyybber so fun discovery, turns out nkError is a problem when the `efNoSemCheck` flag is passed and causes all sorts of issues with the current compiler design. Maybe I'm wrong about the broad reaching impact, but here goes.

The command to make the current PR fail is `./koch.py temp -d:debug c -r tests/stdlib/tsequtils.nim` It'll spew a giant pile of debug on screen, it's the body of `mapIt` from `sequtils` because it's failing in the test on line 432.

What's happening is the `var it {.inject.} = ...` in `mapIt` is producing an nkError, great but then semBlock isn't doing anything with it. So I changed up `semBlock` to do that, and honestly the code makes way more sense from an analysis and production perspective.

The problem is that it's still failing and that's because `semStmtList` in `semstmts` is failing to to produce an nkError, because it's guarded by a check for `efNoSemCheck` flag. The issue now is that we're producing an AST riddled with nkErrors.

There are a bunch of issues but if I keep propogating nkError we're basically doing no semcheck whatsoever until we get up to the callsite that issued the `efNoSemCheck` in the first place. Then there is the issue we can't separate type vs untyped analysis. Can't recover the error because we're not consistent enough with `wrongNode`, our report kinds aren't sufficiently categorized for this.

Thoughts?

Reply:
@zerbina in case this piques your interests or you have some ideas.

Reply:
Hmm, rereading through history and where it's used `efNoSemCheck` might not be as serious as it sounds and it seems to only really matter wrt to governing whether to run after macro call analysis. It might just be as simple as propagating it through? 🤔

Reply:
I haven't tried reproducing the issue yet, but here are some first thoughts:

If I haven't missed a usage site, `efNoSemCheck` is only passed to `semTemplateExpr` and `semMacroExpr` (as you said). The `efNoSemCheck notin flags` test in `semStmtList` currently always evaluates to true (and `isError` should thus be set to true in case an `nkError` node is encountered), since both `semTemplateExpr` and `semMacroExpr` prevent post-expansion semantic analysis if `efNoSemCheck` is present.

Aside: Maybe `efNoSemCheck` should be removed and both `semTemplateExpr` and `semMacroExpr` be split into two procedures, where one only performs the expansion and the other one performs the expansion plus `semAfterMacro`?

Reply:
Thanks, for sharing your perspective, it corroborates what I'm experiencing/thinking.

Right now I'm trying to sort out where the nkError get swallowed -- got it progagating it further. The nkError wedge seems to be pushing in the same direction that you're hinting at, which is to split the expansion vs check.

Reply:
Not pretty, but it seems to work.